substra/sdk/backends/local/backend.py

```diff
@@ -0,0 +1,523 @@
+            # validate test data samples
+            if spec.test_data_sample_keys is None:
+                spec.test_data_sample_keys = list()
+            for key in spec.test_data_sample_keys:
+                self._db.get(schemas.Type.DataSample, key)
+            test_dataset = {
+                "dataset_key": spec.test_data_manager_key,
+                "data_sample_keys": spec.test_data_sample_keys,
+            }
+
+        # create objective model instance
+        objective = models.Objective(
+            key=fs.hash_file(spec.metrics),
+            pkhash=fs.hash_file(spec.metrics),
+            name=spec.name,
+            owner=_BACKEND_ID,
+            test_dataset=test_dataset,
+            permissions={
+                "process": {
+                    "public": permissions.public,
+                    "authorized_ids": permissions.authorized_ids,
+                },
+            },
+            description=str(spec.description),
+            metrics=str(spec.metrics),
+            metadata=spec.metadata if spec.metadata else dict(),
+        )
+
+        # add objective to storage and update optionnally the associated dataset
+        objective = self._db.add(objective, exist_ok)
+        if spec.test_data_manager_key:
+            dataset.objective_key = objective.key
+
+        return objective
+
+    def __add_algo(self, model_class, spec, exist_ok, spec_options=None):
+        permissions = self.__compute_permissions(spec.permissions)
+        algo = model_class(
+            key=fs.hash_file(spec.file),
+            pkhash=fs.hash_file(spec.file),
+            name=spec.name,
+            owner=_BACKEND_ID,
+            permissions={
+                "process": {
+                    "public": permissions.public,
+                    "authorized_ids": permissions.authorized_ids,
+                },
+            },
+            file=str(spec.file),
+            description=str(spec.description),
+            metadata=spec.metadata if spec.metadata else dict(),
+        )
+        return self._db.add(algo, exist_ok)
+
+    def _add_algo(self, spec, exist_ok, spec_options=None):
+        return self.__add_algo(models.Algo, spec, exist_ok, spec_options=spec_options)
+
+    def _add_aggregate_algo(self, spec, exist_ok, spec_options=None):
+        return self.__add_algo(
+            models.AggregateAlgo, spec, exist_ok, spec_options=spec_options
+        )
+
+    def _add_composite_algo(self, spec, exist_ok, spec_options=None):
+        return self.__add_algo(
+            models.CompositeAlgo, spec, exist_ok, spec_options=spec_options
+        )
+
+    def __intersect_permissions(self, aggregate_set, element):
+        if not element.permissions.process.public:
+            if aggregate_set is None:
+                aggregate_set = set(element.permissions.process.authorized_ids)
+            else:
+                aggregate_set = aggregate_set & set(
```

Comment:
Could you use the `intersection` method here instead of `&` ? I find it much more explicit and the doc stresses that it's more readable.

Reply:
Also, this is a static method, maybe it could be moved to an independant function instead of a method?